TF_Service_dotNet/TouchFree/Connections/ConnectionMessageTypes.cs

```diff
@@ -2,149 +2,103 @@
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
 using Ultraleap.TouchFree.Library.Configuration;
+using Ultraleap.TouchFree.Library.Connections.DiagnosticApi;
 
-namespace Ultraleap.TouchFree.Library.Connections
-{
-    public enum TrackingServiceState
-    {
-        UNAVAILABLE,
-        NO_CAMERA,
-        CONNECTED
-    }
+namespace Ultraleap.TouchFree.Library.Connections;
 
-    public enum BinaryMessageType
-    {
-        Hand_Data = 1
-    }
+public enum TrackingServiceState
+{
+    UNAVAILABLE,
+    NO_CAMERA,
+    CONNECTED
+}
 
-    public enum ConfigurationState
-    {
-        NOT_LOADED,
-        LOADED,
-        ERRORED
-    }
+public enum BinaryMessageType
+{
+    Hand_Data = 1
+}
 
-    [Serializable]
-    public struct ConfigState
-    {
-        public string requestID;
-        public InteractionConfig interaction;
-        public PhysicalConfig physical;
+public enum ConfigurationState
+{
+    NOT_LOADED,
+    LOADED,
+    ERRORED
+}
 
-        public ConfigState(string _id, InteractionConfig _interaction, PhysicalConfig _physical)
-        {
-            requestID = _id;
-            interaction = _interaction;
-            physical = _physical;
-        }
-    }
+[Serializable]
+public readonly record struct ConfigState(string requestID, InteractionConfig interaction, PhysicalConfig physical);
+
+[Serializable]
+public readonly record struct ServiceStatus(string requestID,
+    TrackingServiceState trackingServiceState,
+    ConfigurationState configurationState,
+    string serviceVersion,
+    string trackingVersion,
+    string cameraSerial,
+    string cameraFirmwareVersion)
+{
+    public static ServiceStatus FromDApiTypes(string requestId,
+        TrackingServiceState trackingServiceState,
+        ConfigurationState configurationState,
+        string serviceVersion,
+        in ApiInfo? apiInfo,
+        in DeviceInfo? deviceInfo) =>
+        new(requestId,
+            trackingServiceState,
+            configurationState,
+            serviceVersion,
+            apiInfo.GetValueOrDefault().ServiceVersion ?? "Tracking not connected",
+            deviceInfo.GetValueOrDefault().Serial ?? "Device not connected",
+            deviceInfo.GetValueOrDefault().Firmware ?? "Device not connected");
```

Comment:
Will these default to a struct with ServiceVersion etc as null and so fall back to the default error string?

Reply:
Yes it's same behaviour as before - strings in `default<T?>` where T is a struct are null.